#Lab 2: Probability Calculation for Red Dragon Inn

In [2]:
import itertools
import pandas as pd
import numpy as np

**Part A: Sum of dice greater than x**

In [3]:
def rollOptions(m: int, n: int):
  if m < 1 or m > 4 or n < 1 or n > 20: return []
  return list(itertools.product(range(1, n+1), repeat=m))

# chance of rolling mdn > x
# note: in the example code, he used >=, but in the instructions it said >
def rollDice(m: int, n: int, x: int):
  sums = np.array([sum(roll) for roll in rollOptions(m,n)])
  return np.mean(sums > x)

# print(rollDice(3, 6, 5))

**Part B: At least one die greater than or equal to x**

In [4]:
# chance of rolling at least one die >= x
def rollAtLeastOne(m: int, n: int, x: int):
  rolls = rollOptions(m, n)
  return np.mean([any(die >= x for die in roll) for roll in rolls])

# rollAtLeastOne(3, 6, 5)

**Tool Performace Evaluation**

Roll 1d20 and get a result of 17 or more

In [5]:
rollDice(1, 20, 16)

0.2

Roll 3d4 and get a result of 10 or more

In [6]:
rollDice(3, 4, 9)

0.15625

Roll 4d6 and have at least one dice show a 6

In [9]:
rollAtLeastOne(4, 6, 6)

0.5177469135802469

**Summary of probabilities found & best choice**

The probabilites of successful completing the three challenges presented to the player are as follows:


Roll 1d20 and get a result of 17 or more:  **0.2**


Roll 3d4 and get a result of 10 or more: **0.156**


Roll 4d6 and have at least one dice show a 6:  **0.518**

Given these probabilites, the best choice for the player is the third challenge: Roll 4d6 and have at least one dice show a 6. The player has the best chance of successfully completing this challenge.